component/loki/source/heroku/heroku.go

```diff
@@ -0,0 +1,198 @@
+// component.
+type Arguments struct {
+	HerokuListener       ListenerConfig      `river:"listener,block"`
+	Labels               map[string]string   `river:"labels,attr,optional"`
+	UseIncomingTimestamp bool                `river:"use_incoming_timestamp,attr,optional"`
+	ForwardTo            []loki.LogsReceiver `river:"forward_to,attr"`
+	RelabelRules         *flow_relabel.Rules `river:"relabel_rules,attr,optional"`
+}
+
+// ListenerConfig defines a heroku listener.
+type ListenerConfig struct {
+	ListenAddress string `river:"address,attr,optional"`
+	ListenPort    int    `river:"port,attr"`
+	// TODO - add the rest of the server config from Promtail
+}
+
+// DefaultListenerConfig provides the default arguments for a heroku listener.
+var DefaultListenerConfig = ListenerConfig{
+	ListenAddress: "0.0.0.0",
+}
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (lc *ListenerConfig) UnmarshalRiver(f func(interface{}) error) error {
+	*lc = DefaultListenerConfig
+
+	type herokucfg ListenerConfig
+	err := f((*herokucfg)(lc))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Component implements the loki.source.heroku component.
+type Component struct {
+	opts    component.Options
+	metrics *ht.Metrics
+
+	mut    sync.RWMutex
+	lc     ListenerConfig
+	fanout []loki.LogsReceiver
+	target *ht.HerokuTarget
+
+	handler loki.LogsReceiver
+}
+
+// New creates a new loki.source.heroku component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts:    o,
+		metrics: ht.NewMetrics(o.Registerer),
+		mut:     sync.RWMutex{},
+		lc:      ListenerConfig{},
+		fanout:  args.ForwardTo,
+		target:  nil,
+		handler: make(loki.LogsReceiver),
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+
+		level.Info(c.opts.Logger).Log("msg", "loki.source.heroku component shutting down, stopping listener")
+		if c.target != nil {
+			err := c.target.Stop()
+			if err != nil {
+				level.Error(c.opts.Logger).Log("msg", "error while stopping heroku listener", "err", err)
+			}
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			c.mut.RLock()
+			for _, receiver := range c.fanout {
+				receiver <- entry
+			}
+			c.mut.RUnlock()
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	newArgs := args.(Arguments)
+	c.fanout = newArgs.ForwardTo
+
+	var rcs []*relabel.Config
+	if newArgs.RelabelRules != nil && len(newArgs.RelabelRules.GetAll()) > 0 {
+		rcs = flow_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelRules.GetAll())
+	}
+
+	if configsChanged(c.lc, newArgs.HerokuListener) {
```

Comment:
I think this comparison will always trigger as we're never storing a copy of the new config in `c.lc`.

Moreover, we might also have to check whether the relabeling rules have changed; otherwise they will never be propagated to NewHerokuTarget (same goes for the syslog source component, too).